{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from sys import argv\n",
    "import numpy as np\n",
    "generator = SeqIO.parse(\"SP1 copy.fq\", \"fastq\")\n",
    "out_put = open(\"out_file\", \"w+\")\n",
    "trim_out = open(\"temp_file\", \"w+\")\n",
    "quality_seqs = []\n",
    "good_reads = []\n",
    "passed = 0\n",
    "failed = 0\n",
    "\n",
    "def trimmer():\n",
    "    for record in SeqIO.parse(\"SP1 copy.fq\", \"fastq\"):\n",
    "        score = record.letter_annotations['phred_quality']\n",
    "        for i, q in enumerate(score[a:]):\n",
    "            if q >= 30:\n",
    "                good_reads.append(record)\n",
    "                print(\"Saved %s\" % record.seq[i:])\n",
    "            elif q < 30:\n",
    "                a =+ 1\n",
    "                for i, q in enumerate(score[a:]):\n",
    "                    if q >= 30:\n",
    "                        good_reads.append(record[a:])\n",
    "                        print(\"Trimmed %s\" % record[q:])\n",
    "        SeqIO.write(good_reads, out_put,\"fastq\")\n",
    "        \n",
    "trimmer()\n",
    "print(\"%i Passed Quality check\" % passed)\n",
    "\n",
    "def lenght_check():\n",
    "        for record in SeqIO.parse(\"out_put\", \"fastq\"):\n",
    "            if len(record.seq) >= 30:\n",
    "                quality_seqs.append(record)\n",
    "SeqIO.write(quality_seqs, trim_out, \"fastq\")\n",
    "\n",
    "lenght_check()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from sys import argv\n",
    "import numpy as np\n",
    "generator = SeqIO.parse(\"SP1 copy.fq\", \"fastq\")\n",
    "out_put = open(\"out_file\", \"w+\")\n",
    "trim_out = open(\"temp_file\", \"w+\")\n",
    "quality_seqs = []\n",
    "good_reads = []\n",
    "\n",
    "def trimmer():\n",
    "    for record in SeqIO.parse(\"SP1 copy.fq\", \"fastq\"):\n",
    "        score = record.letter_annotations['phred_quality']\n",
    "        for i, q in enumerate(score):\n",
    "            if q >= 30:\n",
    "                good_reads.append(record[i:])\n",
    "                #print(\"Saved %s\" % record[i:])\n",
    "            elif q < 30:\n",
    "                i =+ 1\n",
    "                #for i, q in enumerate(score):\n",
    "                    #if q >= 30:\n",
    "                        #good_reads.append(record[i:])\n",
    "                        #print(\"Trimmed %s\" % record.id[i:])\n",
    "                        #break\n",
    "        SeqIO.write(good_reads, out_put,\"fastq\")\n",
    "        \n",
    "trimmer()\n",
    "#print(\"%i Passed Quality check\" % passed)\n",
    "\n",
    "def lenght_check():\n",
    "        for record in SeqIO.parse(\"out_put\", \"fastq\"):\n",
    "            if len(record.seq) >= 30:\n",
    "                quality_seqs.append(record[i:])\n",
    "SeqIO.write(quality_seqs, trim_out, \"fastq\")\n",
    "\n",
    "lenght_check()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from sys import argv\n",
    "import numpy as np\n",
    "generator = SeqIO.parse(\"SP1 copy.fq\", \"fastq\")\n",
    "out_put = open(\"out_file\", \"w+\")\n",
    "trim_out = open(\"temp_file\", \"w+\")\n",
    "trimmed_seqs = []\n",
    "good_reads = []\n",
    "filtered_seqs = []\n",
    "merged_seqs = []\n",
    "def trimmer():\n",
    "    for record in SeqIO.parse(\"SP1 copy.fq\", \"fastq\"):\n",
    "        ids = set() \n",
    "        i = 0\n",
    "        a = 0\n",
    "        score = record.letter_annotations['phred_quality']\n",
    "        for i, q in enumerate(score[a:]):\n",
    "            if q >= 30 and record.id not in ids:\n",
    "                    ids.add(record.id)\n",
    "                    SeqIO.write(record, trim_out, \"fastq\")\n",
    "                    i =+ 1\n",
    "            elif q < 30:              \n",
    "                a =+ 1\n",
    "                #for i, q in enumerate(score[a:]):\n",
    "                 #   if q >= 30 and record.id not in ids:\n",
    "                  #          ids.add(record.id)\n",
    "                   #         trimmed_seqs.append(record)\n",
    "                    #elif q < 30:\n",
    "                     #   a =+ 1\n",
    "                    \n",
    "                        \n",
    "        merged_seqs.append(good_reads)\n",
    "        merged_seqs.append(trimmed_seqs)\n",
    "        SeqIO.write(merged_seqs, out_put, \"fastq\")\n",
    "trimmer()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "quality_seqs = []\n",
    "def lenght_check():\n",
    "        for record in SeqIO.parse(\"out_file\", \"fastq\"):\n",
    "            if len(record.seq) >= 30:\n",
    "                quality_seqs.append(record)\n",
    "        SeqIO.write(quality_seqs, \"trim_out\", \"fastq\")\n",
    "\n",
    "lenght_check()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for record in SeqIO.parse(\"trim_out\", \"fastq\"):\n",
    "    print(len(record.seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "good_reads = []\n",
    "trimmed_seqs = []\n",
    "merged_seqs = []\n",
    "filtered = open(\"final_file\", \"w+\")\n",
    "out_put = open(\"out_file\", \"w+\")\n",
    "trim_temp = open(\"temp\", \"w+\")\n",
    "def trimmer():\n",
    "    i = 0\n",
    "    a = 0\n",
    "    for record in SeqIO.parse(\"SP1 copy.fq\", \"fastq\"):\n",
    "        score = record.letter_annotations['phred_quality']\n",
    "        ids = set()\n",
    "        for i, q in enumerate(score):\n",
    "            if q >= 30:\n",
    "                if record.id not in ids:\n",
    "                    good_reads.append(record)\n",
    "                i =+ 1\n",
    "            elif q < 30:\n",
    "                a =+ 1\n",
    "                for i, q in enumerate(score[a:]):\n",
    "                    if q >= 30:\n",
    "                        good_reads.append(record[a:])\n",
    "    SeqIO.write(good_reads, out_put, \"fastq\")\n",
    "\n",
    "\n",
    "\n",
    "        \n",
    "trimmer()\n",
    "print(out_put)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from sys import argv\n",
    "import numpy as np\n",
    "generator = SeqIO.parse(\"SP1 copy.fq\", \"fastq\")\n",
    "out_put = open(\"out_file.fq\", \"w+\")\n",
    "trim_out = open(\"temp_file\", \"w+\")\n",
    "trimmed_seqs = []\n",
    "good_reads = []\n",
    "filtered_seqs = []\n",
    "quality_seqs = []\n",
    "merged_seqs = []\n",
    "records = SeqIO.parse(\"SP1 copy.fq\", \"fastq\")\n",
    "trim_record = SeqIO.parse(\"trim_out\", \"fastq\")\n",
    "def trimmer():\n",
    "    index = 0\n",
    "    a = 0\n",
    "    for record in records:\n",
    "        score = record.letter_annotations['phred_quality']\n",
    "        for index, qual in enumerate(score[a:]):\n",
    "            if int(qual) >= int(30):\n",
    "                good_reads.append(record[a:])    \n",
    "                #print(good_reads, record.letter_annotations[\"phred_quality\"][a])\n",
    "                index =+ 1\n",
    "                a = 0\n",
    "                break\n",
    "            elif int(qual) < int(30):\n",
    "                print(\"%s Has Q-Score: %i Needs To Be Trimmed on %s\" % (record.seq[a], score[a], record.id))\n",
    "                a =+ 1\n",
    "                for index, qual in enumerate(score[a:]):\n",
    "                    if int(qual) >= int(30):\n",
    "                        filtered_seqs.append(record[a:])\n",
    "                    elif int(qual) < int(30):\n",
    "                        print(\"%s Did not Pass QC\" % record)\n",
    "                        break\n",
    "                #for rec in filtered_seqs:\n",
    "                 #   quality = rec.letter_annotations['phred_quality']\n",
    "                  #  for i, qual in enumerate(quality[a:]):\n",
    "                   #     if int(qual) >= int(30):\n",
    "                    #        good_reads.append(record[a:])\n",
    "                     #       break\n",
    "                    #print(\"Sequence: %s Trimmed with leading Q-Score: %i on %s\" % (record.seq[a:], score[a], record.id))\n",
    "                #else:\n",
    "                 #   break\n",
    "                    \n",
    "#def lenght_check():\n",
    "        #for record in SeqIO.parse(\"out_file\", \"fastq\"):\n",
    "            #if len(record.seq) >= 30:\n",
    "               # quality_seqs.append(record)\n",
    "        #SeqIO.write(quality_seqs, \"trim_out\", \"fastq\")\n",
    "\n",
    "#lenght_check()    \n",
    "    SeqIO.write(filtered_seqs, trim_out, \"fastq\")\n",
    "    SeqIO.write(good_reads, out_put, \"fastq\")\n",
    "   # print(good_reads)\n",
    "    #print(\"############\")\n",
    "    #print(filtered_seqs)\n",
    "    \n",
    "#def base_iterator():\n",
    " #   for index, qual in enumerate(score):\n",
    "  #      if int(qual) < int(30):\n",
    "   #         a =+ 1\n",
    "    #        base_iterator()\n",
    "     #   if int(qual) >= int(30):\n",
    "      #      print(\"Sequence: %s Trimmed with %s Q-Score %i on %s\" % (record.seq[a:], record.seq[a], score[a], record.id))\n",
    "        \n",
    "trimmer()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from sys import argv\n",
    "import numpy as np\n",
    "good_reads = []\n",
    "notso_good_reads = []\n",
    "filtered_reads = []\n",
    "records = SeqIO.parse(\"SP1 copy.fq\", \"fastq\")\n",
    "out_fq = open(\"out_file.fq\", \"w+\")\n",
    "a = 0\n",
    "index = 0\n",
    "for record in records:\n",
    "    ids = set()\n",
    "    score = record.letter_annotations['phred_quality']\n",
    "    for index, qual in enumerate(score):\n",
    "        if qual >= 30:\n",
    "            if record.id not in ids:\n",
    "                good_reads.append(record)    \n",
    "                #print(good_reads, record.letter_annotations[\"phred_quality\"][a])\n",
    "            index += 1\n",
    "        elif qual < 30:\n",
    "            a += 1 \n",
    "            #for qual in enumerate(score[a:]):\n",
    "             #   if int(qual) >= int(30) and record.id not in ids:\n",
    "              #          notso_good_reads.append(record[a:])\n",
    "               #         SeqIO.write(notso_good_reads, )\n",
    "                #elif int(qual) < int(30) and record.id not in ids:\n",
    "                 #   a += 1\n",
    "                    \n",
    "            \n",
    "                    \n",
    "           \n",
    "SeqIO.write(good_reads, out_fq, \"fastq\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from sys import argv\n",
    "import numpy as np\n",
    "good_reads = []\n",
    "notso_good_reads = []\n",
    "records = SeqIO.parse(\"SP1 copy.fq\", \"fastq\")\n",
    "out_fq = open(\"out_file.fq\", \"w+\")\n",
    "for record in records:\n",
    "    score = record.letter_annotations['phred_quality']\n",
    "    for i, qual in enumerate(score, start = 0):\n",
    "        if qual >= 30:     \n",
    "            good_reads.append(record)\n",
    "            i += 1\n",
    "        elif qual < 30:\n",
    "            a += 1\n",
    "            \n",
    "        #if len(record.seq[a:]) >= int(30):\n",
    "            #notso_good_reads.append(record[a:])\n",
    "\n",
    "                    \n",
    "#good_reads.append(notso_good_reads)            \n",
    "SeqIO.write(good_reads, out_fq, \"fastq\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from sys import argv\n",
    "reads = SeqIO.parse(\"SP1 copy.fq\", \"fastq\")\n",
    "quality_reads = []\n",
    "out_file = open(\"out_file\", \"w+\")\n",
    "\n",
    "def read_ripper():\n",
    "    for record in reads:\n",
    "        i = 0   \n",
    "        quality = record.letter_annotations[\"phred_quality\"]\n",
    "        for reads, q in enumerate(quality):\n",
    "            if q < 30:\n",
    "                i += 1\n",
    "            elif q >= 30:\n",
    "                quality_reads.append(record[i:])\n",
    "            \n",
    "    SeqIO.write(quality_reads, out_file, \"fastq\")\n",
    "read_ripper()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from sys import argv\n",
    "import numpy as np\n",
    "generator = SeqIO.parse(\"SP1 copy.fq\", \"fastq\")\n",
    "out_put = open(\"out_file.fq\", \"w+\")\n",
    "trim_out = open(\"temp_file.fq\", \"w+\")\n",
    "trimmed_seqs = []\n",
    "good_reads = []\n",
    "filtered_seqs = []\n",
    "quality_seqs = []\n",
    "merged_seqs = []\n",
    "records = SeqIO.parse(\"SP1 copy.fq\", \"fastq\")\n",
    "trim_record = SeqIO.parse(\"trim_out\", \"fastq\")\n",
    "\n",
    "def length_filter():\n",
    "    for record in trim_record:\n",
    "        for index, length in enumerate(record.seq, start = 0):\n",
    "            if len(length) >= 30:\n",
    "                good_reads.append(record)\n",
    "            index += 1\n",
    "                \n",
    "            \n",
    "def quality_filter():\n",
    "    for record in records:\n",
    "        index = 0\n",
    "        a = 0\n",
    "        score = record.letter_annotations['phred_quality']\n",
    "        for index, qual in enumerate(score):\n",
    "            if int(qual) >= int(30):\n",
    "                good_reads.append(record)\n",
    "                index += 1\n",
    "                a = 0\n",
    "                break\n",
    "            elif int(qual) < int(30):\n",
    "                a += 1\n",
    "                #for index, qual in enumerate(score[a:]):\n",
    "                if int(qual) >= int(30):\n",
    "                    good_reads.append(record[a:])\n",
    "                    break\n",
    "                   #     good_reads.append(filtered_seqs)\n",
    "                    #elif int(qual) < int(30) or len(record.seq[a:]) < int(30):\n",
    "                     #   a += 1\n",
    "                      #  break\n",
    "   # SeqIO.write(filtered_seqs, trim_out, \"fastq\")\n",
    "    SeqIO.write(good_reads, out_put, \"fastq\")\n",
    "    for record in good_reads:\n",
    "        for index, length in enumerate(record.seq):\n",
    "            if len(record.seq) >= 30:\n",
    "                trimmed_seqs.append(record)\n",
    "    SeqIO.write(trimmed_seqs, trim_out, \"fastq\")\n",
    "        \n",
    "\n",
    "trimmer()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from sys import argv \n",
    "import numpy as np\n",
    "reads = SeqIO.parse(\"SP1 copy.fq\", \"fastq\")\n",
    "trimmed_reads = SeqIO.parse(\"good_reads.fq\", \"fastq\")\n",
    "good_reads = []\n",
    "final_reads = []\n",
    "def read_trimmer():\n",
    "    for record in reads:\n",
    "        a = 0\n",
    "        quality = record.letter_annotations['phred_quality']\n",
    "        for i, q in enumerate(quality):\n",
    "            if int(q) >= int(30):\n",
    "                good_reads.append(record[a:])\n",
    "                break\n",
    "            else:\n",
    "                a += 1\n",
    "    SeqIO.write(good_reads, \"good_reads.fq\", \"fastq\")\n",
    "read_trimmer()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def lenght_check():\n",
    "        for record in trimmed_reads:\n",
    "            if len(record.seq) >= 30:\n",
    "                final_reads.append(record)\n",
    "        SeqIO.write(final_reads, \"trim_out.fq\", \"fastq\")\n",
    "lenght_check()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from sys import argv\n",
    "import os\n",
    "#script, fq_in, fq_out, min_q, min_length = argv\n",
    "reads = SeqIO.parse(\"SP1 copy.fq\", \"fastq\")\n",
    "trimmed_reads = SeqIO.parse(\"good_reads.fq\", \"fastq\")\n",
    "good_reads = []\n",
    "final_reads = []\n",
    "\n",
    "def get_read():\n",
    "    count = 0\n",
    "    for record in reads:\n",
    "        print(\"ID:%s\" % record.id, \"\\nSequence:%s\" % record.seq, \"\\nDescription:%s\" % record.description, \"\\nQuality Score:%s\" % record.letter_annotations[\"phred_quality\"]) \n",
    "        count += 1\n",
    "def read_trimmer():\n",
    "    for record in reads:\n",
    "        a = 0\n",
    "        quality = record.letter_annotations['phred_quality']\n",
    "        for i, q in enumerate(quality):\n",
    "            if int(q) >= int(30):\n",
    "                good_reads.append(record[a:])\n",
    "                break\n",
    "            else:\n",
    "                a += 1\n",
    "    SeqIO.write(good_reads, \"good_reads.fq\", \"fastq\")\n",
    "\n",
    "def lenght_check():\n",
    "    reject = 0\n",
    "    for record in trimmed_reads:\n",
    "        if len(record.seq) >= int(30):\n",
    "            final_reads.append(record)\n",
    "            passed += 1\n",
    "        else:\n",
    "            reject += 1\n",
    "    SeqIO.write(final_reads, \"Trimmed_output.fq\", \"fastq\")\n",
    "    print(\"%i Reads Failed Quality Control\\n%i Trimmed and Written to New File \" % (reject, passed)\n",
    "\n",
    "\n",
    "def main():\n",
    "    read_trimmer()\n",
    "    lenght_check()\n",
    "    os.remove(\"good_reads.fq\")\n",
    "    #print(\"%i\" % reject)\n",
    "main()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from sys import argv\n",
    "import os\n",
    "#script, fq_in, fq_out, min_q, min_length = argv\n",
    "reads = SeqIO.parse(\"SP1 copy.fq\", \"fastq\")\n",
    "trimmed_reads = SeqIO.parse(\"good_reads.fq\", \"fastq\")\n",
    "good_reads = []\n",
    "final_reads = []\n",
    "\n",
    "def get_read():\n",
    "    count = 0\n",
    "    for record in reads:\n",
    "        print(\"ID:%s\" % record.id, \"\\nSequence:%s\" % record.seq, \"\\nDescription:%s\" % record.description, \"\\nQuality Score:%s\" % record.letter_annotations[\"phred_quality\"]) \n",
    "        count += 1\n",
    "    print(\"%i Reads in Original File\" % count)\n",
    "def read_trimmer():\n",
    "    for record in reads:\n",
    "        a = 0\n",
    "        quality = record.letter_annotations['phred_quality']\n",
    "        for i, q in enumerate(quality):\n",
    "            if int(q) >= int(30):\n",
    "                good_reads.append(record[a:])\n",
    "                break\n",
    "            else:\n",
    "                a += 1\n",
    "    SeqIO.write(good_reads, \"good_reads.fq\", \"fastq\")\n",
    "\n",
    "def lenght_check():\n",
    "    reject = 0\n",
    "    for record in trimmed_reads:\n",
    "        if len(record.seq) >= int(30):\n",
    "            final_reads.append(record)\n",
    "            passed += 1\n",
    "        else:\n",
    "            reject += 1\n",
    "    SeqIO.write(final_reads, \"Trimmed_output.fq\", \"fastq\")\n",
    "    print(\"%i Reads Failed Quality Control\\n%i Trimmed and Written to New File \" % (reject, passed)\n",
    "\n",
    "\n",
    "def main():\n",
    "    read_trimmer()\n",
    "    lenght_check()\n",
    "    os.remove(\"good_reads.fq\")\n",
    "main()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from sys import argv\n",
    "import os\n",
    "#script, fq_in, fq_out, min_q, min_length = argv\n",
    "reads = SeqIO.parse(\"SP1 copy.fq\", \"fastq\")\n",
    "trimmed_reads = SeqIO.parse(\"good_reads.fq\", \"fastq\")\n",
    "final_reads = SeqIO.parse(\"razzle_dazzle.fq\", \"fastq\")\n",
    "good_reads = []\n",
    "final_reads = []\n",
    "\n",
    "def get_read():\n",
    "    for record in reads:\n",
    "        print(\"ID:%s\" % record.id, \"\\nSequence:%s\" % record.seq, \"\\nDescription:%s\" % record.description, \"\\nQuality Score:%s\" % record.letter_annotations[\"phred_quality\"]) \n",
    "    \n",
    "def read_trimmer():\n",
    "    for record in reads:\n",
    "        a = 0\n",
    "        quality = record.letter_annotations['phred_quality']\n",
    "        for i, q in enumerate(quality):\n",
    "            if int(q) >= int(30):\n",
    "                good_reads.append(record[a:])\n",
    "                break\n",
    "            else:\n",
    "                a += 1\n",
    "    SeqIO.write(good_reads, \"good_reads.fq\", \"fastq\")\n",
    "\n",
    "def lenght_check():\n",
    "    reject = 0\n",
    "    passed = 0\n",
    "    for record in trimmed_reads:\n",
    "        if len(record.seq) >= int(30):\n",
    "            final_reads.append(record)\n",
    "            passed += 1\n",
    "        else:\n",
    "            reject += 1\n",
    "    SeqIO.write(final_reads, \"razzle_dazzle.fq\", \"fastq\")\n",
    "    print(\"Trimming Complete!\")\n",
    "def report_stats():\n",
    "    orig = 0\n",
    "    passed = 0\n",
    "    for record in reads:\n",
    "        orig += 1\n",
    "    for record in final_reads:\n",
    "        passed += 1\n",
    "    print(\"%i Original Reads\\n%i Trimmed\\n%i Failed Quality Control\" % (orig, passed, (orig - passed))\n",
    "def main():\n",
    "    print(\"Trimming your reads now...\")\n",
    "    get_reads()\n",
    "    read_trimmer()\n",
    "    lenght_checker()\n",
    "    report_stats()\n",
    "          "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from sys import argv\n",
    "import os\n",
    "script, fq_in, fq_out, min_q, min_length = argv\n",
    "reads = SeqIO.parse(fq_in, \"fastq\")\n",
    "trimmed_reads = SeqIO.parse(\"good_reads.fq\", \"fastq\")\n",
    "good_reads = []\n",
    "final_reads = []\n",
    "\n",
    "def get_read():\n",
    "    for record in reads:\n",
    "        print(\"ID:%s\" % record.id, \"\\nSequence:%s\" % record.seq, \"\\nDescription:%s\" % record.description, \"\\nQuality Score:%s\" % record.letter_annotations[\"phred_quality\"]) \n",
    "\n",
    "def read_trimmer():\n",
    "    for record in reads:\n",
    "        a = 0\n",
    "        quality = record.letter_annotations['phred_quality']\n",
    "        for i, q in enumerate(quality):\n",
    "            if int(q) >= int(min_q):\n",
    "                good_reads.append(record[a:])\n",
    "                break\n",
    "            else:\n",
    "                a += 1\n",
    "    SeqIO.write(good_reads, \"good_reads.fq\", \"fastq\")\n",
    "\n",
    "def lenght_check():\n",
    "    reject = 0\n",
    "    for record in trimmed_reads:\n",
    "        if len(record.seq) >= int(min_length):\n",
    "            final_reads.append(record)\n",
    "        else:\n",
    "            reject += 1\n",
    "    SeqIO.write(final_reads, fq_out, \"fastq\")\n",
    "\n",
    "    \n",
    "    \n",
    "def report_stats():\n",
    "    orig = 0\n",
    "    passed = 0\n",
    "    for record in reads:\n",
    "        orig += 1\n",
    "    for record in final_reads:\n",
    "        passed += 1\n",
    "    print(\"%i Original Reads\\n%i Trimmed\\n%i Failed Quality Control\" % (orig, passed, (orig - passed))    \n",
    "\n",
    "\n",
    "def main():\n",
    "    print(\"Trimming your reads now\")\n",
    "    #get_read()\n",
    "    read_trimmer()\n",
    "    lenght_check()\n",
    "    report_stats()\n",
    "    os.remove(\"good_reads.fq\")\n",
    "    \n",
    "main()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading File...\n",
      "Trimming File...\n",
      "240 Reads were found\n",
      "72 Reads were removed\n",
      "168 Reads were trimmed and kept\n",
      "Done.\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "from sys import argv\n",
    "import os\n",
    "#script, fq_in, fq_out, min_q, min_length = argv\n",
    "reads = SeqIO.parse(\"SP1 copy.fq\", \"fastq\")\n",
    "get_read = SeqIO.parse(\"SP1 copy.fq\", \"fastq\")\n",
    "trimmed_reads = SeqIO.parse(\"good_reads.fq\", \"fastq\")\n",
    "good_reads = []\n",
    "final_reads = []\n",
    "\n",
    "\n",
    "def read_trimmer():\n",
    "    for record in reads:\n",
    "        a = 0\n",
    "        quality = record.letter_annotations['phred_quality']\n",
    "        for i, q in enumerate(quality):\n",
    "            if int(q) >= int(30):\n",
    "                good_reads.append(record[a:])\n",
    "                break\n",
    "            else:\n",
    "                a += 1\n",
    "    SeqIO.write(good_reads, \"good_reads.fq\", \"fastq\")\n",
    "\n",
    "def lenght_check():\n",
    "    passed = 0\n",
    "    reject = 0\n",
    "    for record in trimmed_reads:\n",
    "        if len(record.seq) >= int(30):\n",
    "            final_reads.append(record)\n",
    "            passed += 1\n",
    "        else:\n",
    "            reject += 1\n",
    "    SeqIO.write(final_reads, \"happy.fq\", \"fastq\")\n",
    "    print(\"%i Reads were found\\n%i Reads were removed\\n%i Reads were trimmed and kept\" % ((passed + reject), reject, passed))\n",
    "\n",
    "def main():\n",
    "    print(\"Reading File...\")\n",
    "    read_trimmer()\n",
    "    print(\"Trimming File...\")\n",
    "    lenght_check()\n",
    "    os.remove(\"good_reads.fq\")\n",
    "    print(\"Done.\")\n",
    "main()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from sys import argv\n",
    "import os\n",
    "script, fq_in, fq_out, min_q, min_length = argv\n",
    "reads = SeqIO.parse(fq_in, \"fastq\")\n",
    "get_read = SeqIO.parse(fq_in, \"fastq\")\n",
    "trimmed_reads = SeqIO.parse(\"good_reads.fq\", \"fastq\")\n",
    "good_reads = []\n",
    "final_reads = []\n",
    "\n",
    "\"\"\"A Python script that performs simple read trimming of a FASTQ file.\n",
    "\n",
    ".. module:: Project01\n",
    "   :platform: Unix, Windows\n",
    "   :synopsis: This script receives as input a FASTQ file and a set of arguments\n",
    "      that control trimming. A new FASTQ file is generated containing only\n",
    "      trimmed reads that meet the given requirements.\n",
    "\n",
    ".. moduleauthor:: Andrew J. Lefors\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "def get_read(fq_in):\n",
    "    \"\"\"Extract a single read from a FASTQ file.\n",
    "\n",
    "    Reads in a FASTQ file are stored in 4 lines that contain the\n",
    "    sequence_id, nucleotide sequence, a second id, and a series of\n",
    "    characters represeting quality scores.\n",
    "\n",
    "    :param fq: A file handle for the FASTQ file\n",
    "    :type fq: An io.TextIOBase object (created using the open() function).\n",
    "\n",
    "    :return: a list containing 4 strings: the sequence ID, nucleotide sequence,\n",
    "        second ID, and the quality score sequence. If there are no more\n",
    "        sequences in the FASTQ file then this function returns False.\n",
    "    :rtype: a list with four elements\n",
    "    \"\"\"\n",
    "    for record in get_read:\n",
    "        print(f\"Openining {fq_in} for reading...\")\n",
    "        print(\"ID:%s\" % record.id, \"\\nSequence:%s\" % record.seq, \"\\nDescription:%s\" % record.description, \"\\nQuality Score:%s\" % record.letter_annotations[\"phred_quality\"])\n",
    "        break\n",
    "def trim_read_front(fq_in, min_q, min_length):\n",
    "    \"\"\"Trims the low quality nucleotides from the front of a reads' sequences.\n",
    "\n",
    "    This function examines the quality score of each nucleotide sequence\n",
    "    starting from the first position of the sequence. When it encouters a\n",
    "    high quality score it stops trimming and returns an updated read.\n",
    "\n",
    "    :param read: A list containing for elements in this order: the sequence ID,\n",
    "        the nucleotide sequence string, a secondary identifier string, and a\n",
    "        quality score string.\n",
    "    :type read: a list\n",
    "\n",
    "    :param min_q:  The minimum quality score that a nucleotide must have to\n",
    "        not be trimmed (e.g. 20).\n",
    "    :type min_q:  integer\n",
    "\n",
    "    :param min_size:  The minimum size that the sequence must have after\n",
    "        trimming to keep the read (e.g. 25).\n",
    "    :type min_size: integer\n",
    "\n",
    "    :return: a list just like the the get_read() function returns but with the\n",
    "       low-quality reads (and corresponding quality scores) trimmed off the\n",
    "       front of the string. If the remaining trimmed read is smaller than the\n",
    "       desired minimum read length then this function returns False.\n",
    "    :rtype: a list with four elements\n",
    "    \"\"\"\n",
    "    #This section trims reads if leading nucleotide quality is less than min_q\n",
    "    for record in reads:\n",
    "        a = 0\n",
    "        quality = record.letter_annotations['phred_quality']\n",
    "        for i, q in enumerate(quality):\n",
    "            if int(q) >= int(min_q):\n",
    "                good_reads.append(record[a:])\n",
    "                break\n",
    "            else:\n",
    "                a += 1\n",
    "    SeqIO.write(good_reads, \"tmp\", \"fastq\")\n",
    "\n",
    "    #This section filters trimmed reads based on sequence length\n",
    "    passed = 0\n",
    "    reject = 0\n",
    "    for record in trimmed_reads:\n",
    "        if len(record.seq) >= int(min_length):\n",
    "            final_reads.append(record)\n",
    "            passed += 1\n",
    "        else:\n",
    "            reject += 1\n",
    "    print(f\"Openining {fq_out} for writing...\")\n",
    "    SeqIO.write(final_reads, fq_out, \"fastq\")\n",
    "    os.remove(\"tmp\")\n",
    "    print(\"%i Reads were found\\n%i Reads were removed\\n%i Reads were trimmed and kept\" % ((passed + reject), reject, passed))\n",
    "\n",
    "\n",
    "#\n",
    "# The main function for the script.\n",
    "#\n",
    "def main(argv):\n",
    "    \n",
    "    \"\"\"The main function of this script.\n",
    "\n",
    "    After trimming is completed, the fucntion prints out three status lines. The\n",
    "    first indicates the total number of reads that were found. The second\n",
    "    indicates how many reads were removed for being too short after trimming and\n",
    "    the third indicates how many reads were trimmed and kept.\n",
    "\n",
    "    The script will create a new FASTQ file of just the trimmed reads and name\n",
    "    it according to the argument provided by the user when running the script.\n",
    "\n",
    "    :param argv:  The incoming arguments to this script as\n",
    "       provided by the sys.argv variable.  There must be four total arguments\n",
    "       provided to the script must be in the following order\n",
    "\n",
    "       - The filename for the input FASTQ file\n",
    "       - The filename for the new output FASTQ file that this script creates\n",
    "       - An integer for the minimum quality score. Anything below this at the\n",
    "         beginning of each read's nucleotide sequence is trimmed off.\n",
    "       - An integer indicating how large a read's nucleotide sequence must\n",
    "         be after trimming in order to keep it.\n",
    "    \"\"\"\n",
    "    get_read()\n",
    "    trim_read_front()\n",
    "#This section counts the number of passed and failed reads and reports it to std.out\n",
    "    orig = 0\n",
    "    passed = 0\n",
    "    for record in reads:\n",
    "        orig += 1\n",
    "    for record in final_reads:\n",
    "        passed += 1\n",
    "    print(\"%i Original Reads\\n%i Trimmed\\n%i Failed Quality Control\" % (orig, passed, (orig - passed))\n",
    "\n",
    "# Begin the program by calling the main function.\n",
    "\n",
    "\n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "unexpected EOF while parsing (<ipython-input-3-2ddbabd271ca>, line 136)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-3-2ddbabd271ca>\"\u001b[0;36m, line \u001b[0;32m136\u001b[0m\n\u001b[0;31m    #main(argv)\u001b[0m\n\u001b[0m               ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m unexpected EOF while parsing\n"
     ]
    }
   ],
   "source": [
    "\"\"\"A Python script that performs simple read trimming of a FASTQ file.\n",
    "\n",
    ".. module:: Project01\n",
    "   :platform: Unix, Windows\n",
    "   :synopsis: This script receives as input a FASTQ file and a set of arguments\n",
    "      that control trimming. A new FASTQ file is generated containing only\n",
    "      trimmed reads that meet the given requirements.\n",
    "\n",
    ".. moduleauthor:: ENTER YOUR NAME HERE\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "from Bio import SeqIO\n",
    "from sys import argv\n",
    "import os\n",
    "script, fq_in, fq_out, min_q, min_length = argv\n",
    "reads = SeqIO.parse(fq_in, \"fastq\")\n",
    "get_read = SeqIO.parse(fq_in, \"fastq\")\n",
    "trimmed_reads = SeqIO.parse(\"good_reads.fq\", \"fastq\")\n",
    "good_reads = []\n",
    "final_reads = []\n",
    "\n",
    "def get_read(get_read):\n",
    "    \"\"\"Extract a single read from a FASTQ file.\n",
    "\n",
    "    Reads in a FASTQ file are stored in 4 lines that contain the\n",
    "    sequence_id, nucleotide sequence, a second id, and a series of\n",
    "    characters represeting quality scores.\n",
    "\n",
    "    :param fq: A file handle for the FASTQ file\n",
    "    :type fq: An io.TextIOBase object (created using the open() function).\n",
    "\n",
    "    :return: a list containing 4 strings: the sequence ID, nucleotide sequence,\n",
    "        second ID, and the quality score sequence. If there are no more\n",
    "        sequences in the FASTQ file then this function returns False.\n",
    "    :rtype: a list with four elements\n",
    "    \"\"\"\n",
    "    for record in get_read:\n",
    "          print(f\"Openining {fq_in} for reading...\")\n",
    "          print(\"ID:%s\" % record.id, \"\\nSequence:%s\" % record.seq, \"\\nDescription:%s\" % record.description, \"\\nQuality Score:%s\" % record.letter_annotations[\"phred_quality\"])\n",
    "          break\n",
    "\n",
    "\n",
    "def trim_read_front(reads, min_q, min_length):\n",
    "    \"\"\"Trims the low quality nucleotides from the front of a reads' sequences.\n",
    "\n",
    "    This function examines the quality score of each nucleotide sequence\n",
    "    starting from the first position of the sequence. When it encouters a\n",
    "    high quality score it stops trimming and returns an updated read.\n",
    "\n",
    "    :param read: A list containing for elements in this order: the sequence ID,\n",
    "        the nucleotide sequence string, a secondary identifier string, and a\n",
    "        quality score string.\n",
    "    :type read: a list\n",
    "\n",
    "    :param min_q:  The minimum quality score that a nucleotide must have to\n",
    "        not be trimmed (e.g. 20).\n",
    "    :type min_q:  integer\n",
    "\n",
    "    :param min_size:  The minimum size that the sequence must have after\n",
    "        trimming to keep the read (e.g. 25).\n",
    "    :type min_size: integer\n",
    "\n",
    "    :return: a list just like the the get_read() function returns but with the\n",
    "       low-quality reads (and corresponding quality scores) trimmed off the\n",
    "       front of the string. If the remaining trimmed read is smaller than the\n",
    "       desired minimum read length then this function returns False.\n",
    "    :rtype: a list with four elements\n",
    "    \"\"\"\n",
    "    for record in reads:\n",
    "        a = 0\n",
    "        quality = record.letter_annotations['phred_quality']\n",
    "        for i, q in enumerate(quality):\n",
    "            if int(q) >= int(min_q):\n",
    "                good_reads.append(record[a:])\n",
    "                break\n",
    "            else:\n",
    "                a += 1\n",
    "    SeqIO.write(good_reads, \"tmp\", \"fastq\")\n",
    "\n",
    "      #This section filters trimmed reads based on sequence length\n",
    "    passed = 0\n",
    "    reject = 0\n",
    "    for record in trimmed_reads:\n",
    "        if len(record.seq) >= int(min_length):\n",
    "            final_reads.append(record)\n",
    "            passed += 1\n",
    "        else:\n",
    "            reject += 1\n",
    "    print(f\"Openining {fq_out} for writing...\")\n",
    "    SeqIO.write(final_reads, fq_out, \"fastq\")\n",
    "    os.remove(\"tmp\")\n",
    "    print(\"%i Reads were found\\n%i Reads were removed\\n%i Reads were trimmed and kept\" % ((passed + reject), reject, passed))\n",
    "\n",
    "\n",
    "\n",
    "#\n",
    "# The main function for the script.\n",
    "#\n",
    "#def main(argv):\n",
    "    \"\"\"The main function of this script.\n",
    "\n",
    "    After trimming is completed, the fucntion prints out three status lines. The\n",
    "    first indicates the total number of reads that were found. The second\n",
    "    indicates how many reads were removed for being too short after trimming and\n",
    "    the third indicates how many reas were trimmed and kept.\n",
    "\n",
    "    The script will create a new FASTQ file of just the trimmed reads and name\n",
    "    it according to the argument provided by the user when running the script.\n",
    "\n",
    "    :param argv:  The incoming arguments to this script as\n",
    "       provided by the sys.argv variable.  There must be four total arguments\n",
    "       provided to the script must be in the following order\n",
    "\n",
    "       - The filename for the input FASTQ file\n",
    "       - The filename for the new output FASTQ file that this script creates\n",
    "       - An integer for the minimum quality score. Anything below this at the\n",
    "         beginning of each read's nucleotide sequence is trimmed off.\n",
    "       - An integer indicating how large a read's nucleotide sequence must\n",
    "         be after trimming in order to keep it.\n",
    "    \"\"\"\n",
    "get_read(reads)\n",
    "trim_read_front(reads, min_q, min_length)\n",
    "#This section counts the number of passed and failed reads and reports it to std.out\n",
    "orig = 0\n",
    "passed = 0\n",
    "for record in reads:\n",
    "    orig += 1\n",
    "for record in final_reads:\n",
    "    passed += 1\n",
    "print(\"%i Original Reads\\n%i Trimmed\\n%i Failed Quality Control\" % (orig, passed, (orig - passed))\n",
    "\n",
    "\n",
    "\n",
    "# Begin the program by calling the main function.\n",
    "#main(argv)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
